jvm-packages/pom.xml

```diff
@@ -34,7 +34,7 @@
         <maven.compiler.source>1.7</maven.compiler.source>
         <maven.compiler.target>1.7</maven.compiler.target>
         <flink.version>1.5.0</flink.version>
-        <spark.version>2.3.3</spark.version>
+        <spark.version>2.4.1</spark.version>
```

Comment:
@CodingCat Is it still possible to compile XGBoost4J with Spark 2.3.x with this change?

Reply:
if you skip tests, yes, 

if you want to run the test, no, because the API in https://github.com/dmlc/xgboost/pull/4377/files#diff-158851aba72186f53a0ff66dd4a792c8R264 does not exist in Spark 2.3, and we have to add in Spark 2.4 to ensure that Spark can handle NaN as expected

Reply:
@CodingCat Is there a way to mask tests not compatible with Spark 2.3.x? I don't think we want to suddenly drop support for Spark 2.3.x

Reply:
I don't think we can mask a single test. The way to skip test is to do something like `mvn package -DskipTests` which skips the compilation of test classes. Once you compile them, e.g. with `mvn package -Dspark.version=2.3.3`, it will fall into cannot find symbol error (since the API does not exist)

According to @srowen from Databricks, the community is going to drop 2.3.x soon, my personal thinking is to move forward (as we have stay in 2.3.x for at least two versions, further, we are going to bring some breaking changes https://github.com/dmlc/xgboost/pull/4349)

Reply:
@CodingCat So I should just use Spark 2.4.1 in the new CI pipeline (will file PR today)?

Reply:
yes, thanks

Reply:
Nice, one fewer things to worry about

Reply:
Yes, Spark 2.3.x will be technically EOL in August. If you want my opinion: I'd keep it simple and forego Spark compatibility with 2.3 if that helps.

That said, what's the runtime issue? maybe I have an easier workaround. There _shouldn't_ be breaking changes between 2.3 and 2.4 (not in the public API) but it is possible either by mistake or exception. If it's that a deprecated method has gone away, maybe there are alternatives.

Reply:
@srowen Mainly the issue is that one of the test being added in this PR will not be compatible with Spark 2.3.x. As @CodingCat said, it is not possible to exclude a single test when running `mvn package`.

Reply:
Can the test simply check the Spark version and skip it if it's not Spark 2.3? IIRC you can check `org.apache.spark.SPARK_VERSION` for example.

Reply:
... or if it's actually a compile-time incompatibility (what is it?) you _could_ make a separate test source tree that is only activated by a profile or specific spark.version in the build. Not sure if it's worth it or what.

Reply:
it's a compilation-time change this line in https://github.com/dmlc/xgboost/pull/4377/files#diff-158851aba72186f53a0ff66dd4a792c8R264 just doesn't compile in 2.3 

and for breaking changes in 2.3-2.4, the parse() method https://github.com/dmlc/xgboost/blob/master/jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/params/DefaultXGBoostParamsWriter.scala#L73 in jackson is brokenly changed in two versions depended by 2.3/2.4....

I don't think 2.3 is that important to us given it's EOL soon

Reply:
OK sounds good. You could work around the first one with reflection maybe, but working around Jackson differences is harder. Well you have my vote for just making the next version focus on Spark 2.4.